modules/data.land/R/extract_soil_nc.R

```diff
@@ -4,91 +4,191 @@
+    }, error = function(e) {
+      xml_doc <- XML::xmlParse(temp_file)
+      all_text <- XML::xpathSApply(xml_doc, "//text()", XML::xmlValue)
+      mukey_candidates <- all_text[grepl("^[0-9,]+$", all_text)]
+      if (length(mukey_candidates) > 0) mukey_candidates[1] else NULL
+    })
+    if (file.exists(temp_file)) unlink(temp_file)
+    
+    mukeys <- strsplit(mukey_str, ",")[[1]]
+    if (length(mukeys) == 0) next
+    
+    mukeys_all <- c(mukeys_all, mukeys)
   }
-
+  
+  # Get unique mukeys from all grid points
+  mukeys_all <- unique(mukeys_all)
+  if (length(mukeys_all) == 0) {
+    PEcAn.logger::logger.severe("No mapunit keys were found for this site.")
+    return(NULL)
+  }
+  
   # calling the query function sending the mapunit keys
   soilprop <- gSSURGO.Query(
-    mukeys,
+    mukeys_all,
     c("chorizon.sandtotal_r",
       "chorizon.silttotal_r",
       "chorizon.claytotal_r",
-      "chorizon.hzdept_r"))
-
-  soilprop.new <- soilprop %>%
+      "chorizon.hzdept_r",
+      "chorizon.hzdepb_r",  
+      "chorizon.om_r",      
+      "chorizon.dbthirdbar_r",  # bulk density at 1/3 bar (field capacity);which is the standard field capacity bulk density measurement
+      "component.comppct_r")) 
+  
+  # Area-weighted aggregation by mukey and horizon depth
+  soilprop.weighted <- soilprop %>%
+    dplyr::mutate(
+      sandtotal_r = as.numeric(sandtotal_r),
+      silttotal_r = as.numeric(silttotal_r),
+      claytotal_r = as.numeric(claytotal_r),
+      hzdept_r = as.numeric(hzdept_r),
+      hzdepb_r = as.numeric(hzdepb_r),
+      om_r = as.numeric(om_r),
+      dbthirdbar_r = as.numeric(dbthirdbar_r),
+      comppct_r = as.numeric(comppct_r)
+    ) %>%
+    dplyr::group_by(mukey, hzdept_r, hzdepb_r) %>%
+    dplyr::summarise(
+      sandtotal_r = weighted.mean(sandtotal_r, comppct_r, na.rm = TRUE),
+      silttotal_r = weighted.mean(silttotal_r, comppct_r, na.rm = TRUE),
+      claytotal_r = weighted.mean(claytotal_r, comppct_r, na.rm = TRUE),
+      om_r = weighted.mean(om_r, comppct_r, na.rm = TRUE),
+      dbthirdbar_r = weighted.mean(dbthirdbar_r, comppct_r, na.rm = TRUE),
+      .groups = "drop"
+    )
+  
+  soilprop.new <- soilprop.weighted %>%
     dplyr::arrange(.data$hzdept_r) %>%
     dplyr::select(
       fraction_of_sand_in_soil = "sandtotal_r",
       fraction_of_silt_in_soil = "silttotal_r",
       fraction_of_clay_in_soil = "claytotal_r",
       soil_depth = "hzdept_r",
+      soil_depth_bottom = "hzdepb_r",
+      organic_matter_pct = "om_r",
+      bulk_density = "dbthirdbar_r",
       mukey = "mukey") %>%
-    dplyr::mutate(dplyr::across(
-        c(dplyr::starts_with("fraction_of"),
-          "soil_depth"),
-        function(x) x / 100))
-
-  soilprop.new <- soilprop.new[ stats::complete.cases(soilprop.new) , ]
+    dplyr::mutate(
+      dplyr::across(c(dplyr::starts_with("fraction_of"), "soil_depth", "soil_depth_bottom"), 
+                    ~ as.numeric(.) / 100),
```

Comment:
As above, we already made sure these are numeric. Coercing them multiple times just makes the pipeline harder to read.

Reply:
👍

Reply:
@divine7022 were you going to change this? Maybe to 


```suggestion
                    ~ . / 100),
```

?